File: examples/from-sql-to-arcflow/01-shared-coworkers.py
"""
Question 1: "Who has worked at the same company as Alice?"

This is the simplest graph traversal — and it's the question that
shows the SQL-to-graph translation cost most clearly. In SQL, you
self-join the employment table. In Cypher, you walk the edge.
"""

from _common import make_duckdb, make_arcflow

print("=" * 68)
print("Q1: Who has worked at the same company as Alice?")
print("=" * 68)

# ──────────────────────── DuckDB / SQL ────────────────────────
print("\n--- DuckDB / SQL ---")
sql = """
SELECT DISTINCT p2.name, o.name AS org
FROM persons p1
JOIN employment e1 ON p1.id = e1.person_id
JOIN orgs o ON e1.org_id = o.id
JOIN employment e2 ON o.id = e2.org_id
JOIN persons p2 ON e2.person_id = p2.id
WHERE p1.name = 'Alice' AND p2.name <> 'Alice'
ORDER BY p2.name, o.name
"""
print(sql.strip())
print("\nresult:")
db = make_duckdb()
for r in db.execute(sql).fetchall():
    print(f"  {r}")
db.close()

# ──────────────────────── ArcFlow / Cypher ────────────────────────
print("\n--- ArcFlow / Cypher ---")
cypher = """
MATCH (alice:Person {name: 'Alice'})-[:EMPLOYED_AT]->(o:Org)<-[:EMPLOYED_AT]-(other:Person)
WHERE other.name <> 'Alice'
RETURN DISTINCT other.name AS name, o.name AS org
ORDER BY name, org
"""
print(cypher.strip())
print("\nresult:")
af = make_arcflow()
for r in af.execute(cypher):
    print(f"  ({r['name']!r}, {r['org']!r})")
af.close()

print("\n--- Observation ---")
print("SQL needs FOUR joins to walk the same edge twice. Cypher walks it once.")
print("Both produce the same result, but as the path length grows, SQL's")
print("self-join chain grows quadratically; Cypher stays one MATCH per hop.")

File: examples/from-sql-to-arcflow/rust/src/bin/01-shared-coworkers.rs
//! Step 01 (Rust) — shared coworkers, both engines side-by-side.
//!
//! Mirrors `01-shared-coworkers.py`. The graph form walks one edge;
//! the SQL form needs four self-joins.

use anyhow::Result;
use from_sql_to_arcflow::load_both;

fn main() -> Result<()> {
    let both = load_both()?;

    println!("=== Shared coworkers of Alice ===\n");

    println!("--- SQL (DuckDB) ---");
    let sql_q = "
        SELECT DISTINCT p2.name AS coworker
        FROM employment e1
        JOIN persons p1 ON p1.id = e1.person_id
        JOIN employment e2 ON e2.org_id = e1.org_id
        JOIN persons p2 ON p2.id = e2.person_id
        WHERE p1.name = 'Alice' AND p2.name != 'Alice'
        ORDER BY p2.name
    ";
    let mut stmt = both.sql.prepare(sql_q)?;
    let rows: Vec<String> = stmt
        .query_map([], |row| row.get::<_, String>(0))?
        .filter_map(|r| r.ok())
        .collect();
    println!("{}", sql_q.trim());
    println!("→ {rows:?}\n");

    println!("--- WorldCypher (ArcFlow) ---");
    let cy = "MATCH (p1:Person {name: 'Alice'})-[:WORKED_AT]->(o:Org)<-[:WORKED_AT]-(p2:Person) \
              WHERE p2.name <> 'Alice' \
              RETURN DISTINCT p2.name AS coworker \
              ORDER BY coworker";
    println!("{cy}");
    let result = both.graph.execute(cy)?;
    let names: Vec<String> = result.rows.into_iter()
        .filter_map(|r| r.get("coworker").cloned())
        .collect();
    println!("→ {names:?}\n");

    println!("Verdict: Cypher walks one edge; SQL needs four self-joins.");
    Ok(())
}
